### DISCUSSION TITLE
Enhancing security insights for RubyGems

### DISCUSSION THREAD

**Discussion by lirantal** [rubygems/rubygems.org#2791] [Ideas]
I'm having a problem or would like to suggest a feature.

This issue is related to:

  - [x] Installing a library
  - [x] The command line `gem`
  - [x] RubyGems.org

---

Hey folks,
Liran from [Snyk](https://snyk.io) (open source security) here 👋

Me and my team have been spending some time recently working with the community to increase the security awareness for developers as they use dependencies. A good couple of examples of that which we've done already with community folks are a [small widget on jsDelivr](https://www.jsdelivr.com/package/npm/lodash?version=4.17.0) and CDNjs [like this](https://cdnjs.dev/libraries/lodash.js/4.10.0).

I'd be keen on exploring some ideas I have about integrating a quick widget on package pages, or even further enhancing the gem/bundler CLI with vulnerability insights on install, much like `npm install` does by default. The RubyGems vulnerabilities database (and other ecosystems) is already open https://snyk.io/vuln?type=rubygems and we could make APIs available to tie all of this up.

Happy to jump on a call or discuss further here/elsewhere if this is interesting for the community.

![image](https://user-images.githubusercontent.com/316371/89864178-2afc5e80-dbb4-11ea-9e24-961218144164.png)

or [like this on yarnpkg](https://yarnpkg.com/package/markdown):

![image](https://user-images.githubusercontent.com/316371/89864434-9f370200-dbb4-11ea-941a-3a69d4eae8ae.png)

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

**Comment by matkoniecz** [comment]
> even further enhancing the gem/bundler CLI with vulnerability insights on install, much like npm install does by default

Note that this would be not enhancement - it just leads to a pointless confusion such as in https://github.com/openstreetmap/iD/issues/5463#issuecomment-435588890

> Yeah - People don't really like this security audit feature because it highlights stuff like this in the dev dependencies that are not real vulnerabilities.

> Even though mapillary is using brfs to assemble the sprites for browserify, the vulnerability can't affect us downstream because we only use the svgs from their repository. iD doesn't actually run brfs anywhere.

> The overagressive warning is reported here: npm/npm#20564 and I think they are working on it, but it's hard to tell since they have switched the project to a new repository.

Maybe it is possible to do it well, but please do not base anything on anything done by npm.

**Comment by lirantal** [comment]
@matkoniecz great context. I'm not here to praise snyk vs npm audit but there's definitely a lot of room for improvement on npm's part that indeed causes a lot of the fatigue you referenced, for example:
* snyk *by default* doesn't report on devDependencies
* snyk's recommended fixes are always the most minimal semver upgrade possible, unlike npm which just updates to *latest* and may cause deps to break due to semver major releases

I'd like to see where we can help with our APIs and vulnerability dataset, whether it is for the CLI tools or for the website.

**Comment by colby-swandale** [comment]
The rubygems team would be happy to hear about any ideas/proposals ideas you have, but please be understandable that we are very cautious about integrating with any 3rd party systems and have a strong preference for managing data/functionality ourselves.

**Comment by matkoniecz** [comment]
Maybe it would be better first as a separate CLI tool allowing to evaluate whatever it is useful at all? 

In addition to npm case - similar attempt by Github keeps spamming my inbox about irrelevant supposed security issues (things that are problem solely in software accepting user input in software that never accept user input from untrusted users and so on).

**Comment by lirantal** [comment]
With regards to testing out the capabilities you can already do that in various ways, and snyk is free for open source devs and maintainers:
* use the snyk cli with [install instructions here](https://support.snyk.io/hc/en-us/articles/360003812538#UUID-a5dbd05f-09c5-b673-4521-ab2c96e453b8)
* or connect simply login to snyk with your github/bitbucket and import a Ruby project.

More about [snyk ruby support over here](https://support.snyk.io/hc/en-us/articles/360006968697-Snyk-for-Ruby).
I'd love to hear your feedback and what you make of it @matkoniecz among the rest of the tooling for ruby in this space.

@colby-swandale totally understood, and makes sense too. we're happy to provide you with any of the APIs you need to get security badges up on rubygems.org for packages, or to integrate with the CLI for more integrated feel. I'd say further more, we could push a vulnerabilities feed down your end of the infra and then you just integrate everything internally without an actual 3rd party API call on the website or the CLI.

We genuinely want to help the community and it's more than just helping out with the vulnerability database and also extends to our security analysts. We handle disclosures and report about ruby gems security issues today and would be happy to help where we can and connect ruby's security team with snyk's for anywhere we can help out, whether it is triaging, providing fixes, or alerting rubygems.org security folks or the gem's maintainers directly if needed for an early heads up.

**Comment by matkoniecz** [comment]
I have not found reasonable way to install it, but I noticed mandatory `snyk auth` ("Once installed, you need to authenticate with your Snyk account:").

I hope that such requirement to authenticate with Snyk would be gone from anything appearing in rubygems.org software, especially things enabled by default.

**Comment by lirantal** [comment]
Good feedback. What would be an ideal way for ruby developers to consume the snyk CLI? I assume a gem? what would be the next best thing? :-)
Just to note that I'm not proposing for any ruby developer to consume the snyk CLI at all, but rather to build these capabilities to already existing ruby CLIs like gem or bundler through APIs and a vulnerability data set that we can provide the community.

Needless to say, a further 'auth' type of things won't be necessary. The relevant ruby CLI could hit an anonymous endpoint on our end, or we'll push the data to rubygems own system and it will query its own, whichever suits best really.

**Comment by matkoniecz** [comment]
> What would be an ideal way for ruby developers to consume the snyk CLI? I assume a gem?

Yes, when I am installing ruby tool I expect `gem install --user-install whatever` similarly like for Python tools I expect `pip` and I guess that node developers expect `npm`. Preferably without need to install system packages and other additional third party tools.

**Comment by lirantal** [comment]
Would using `brew` or `scoop` depending on your OS, or even via a docker container be a reasonable way to consume the tool or do you feel that is too far out there?

**Comment by matkoniecz** [comment]
> Would using brew or scoop depending on your OS, or even via a docker container be a reasonable way to consume the tool or do you feel that is too far out there?

I use solely Linux so it is hard for me to comment what is preferred/normal/used among Windows/Apple users (and even for Linux it is strongly personal opinion). Though I consider Docker containers to be quite weird overkill, useful in cases where someone build extremely complex tool. What sometimes is needed but from description this tool seems like a simple blacklist of insecure versions.

**Comment by lirantal** [comment]
Completely agree. 
I want to circle back to the original point - I think you're totally right about being able to use snyk if it is provided as a ruby package that is easily installed from the ruby toolchain.

Also a friendly ping to @colby-swandale @hsbt.
What would be an ideal way to continue the conversation :-)
I shared some ideas of what can be done and I want to make it the most convenient developer experience for Ruby devs.
